src/utils/mempool.rs

```diff
@@ -137,3 +145,51 @@ impl io::Seek for MemPool {
         io::Seek::seek(&mut self.file, pos)
     }
 }
+
+fn create_shm_fd() -> io::Result<RawFd> {
+    loop {
+        match memfd::memfd_create(
+            CStr::from_bytes_with_nul(b"smithay-client-toolkit\0").unwrap(),
+            memfd::MemFdCreateFlag::MFD_CLOEXEC,
+        ) {
+            Ok(fd) => return Ok(fd),
+            Err(nix::Error::Sys(Errno::EINTR)) => continue,
+            Err(nix::Error::Sys(errno)) => return Err(io::Error::from(errno)),
+            Err(_) => break,
+        }
+    }
+
+    // Fallback to using shm_open
+    let mut rng = thread_rng();
+    let mut mem_file_handle = format!(
+        "/smithay-client-toolkit-{}",
+        rng.gen_range(0, ::std::u32::MAX)
+    );
+    loop {
+        match mman::shm_open(
+            mem_file_handle.as_str(),
+            fcntl::OFlag::O_CREAT
+                | fcntl::OFlag::O_EXCL
+                | fcntl::OFlag::O_RDWR
+                | fcntl::OFlag::O_CLOEXEC,
+            stat::Mode::S_IRUSR | stat::Mode::S_IWUSR,
+        ) {
+            Ok(fd) => match mman::shm_unlink(mem_file_handle.as_str()) {
+                Ok(_) => return Ok(fd),
+                Err(nix::Error::Sys(errno)) => return Err(io::Error::from(errno)),
```

Comment:
Here we need to close the fd, or we'll leak it.

Reply:
Like with [unistd::close](https://docs.rs/nix/0.11.0/nix/unistd/fn.close.html)? Does shm_unlink not do this?

Reply:
yes, using `close()`. `shm_unlink` removes the link of the SHM to the global namespace, preventing other processes from accessing it, but does not close the file descriptor, according to the man page:

> The operation of shm_unlink() is analogous to unlink(2): it removes a shared memory object name, and, once all processes have unmapped  the  object, de-allocates  and destroys the contents of the associated memory region.

Reply:
Closing the fd leads to a bad file descriptor error, maybe its the close or maybe its the wrapping of the fd into a file which closes the fd on drop.

Reply:
Where is it wrapped into a `File`? I just meant closing it in the `Err(nix::Error::Sys(errno))` branch of the `shm_unlink` match, and this branch only, as in this case we do not return the FD from the function...

Reply:
Right, I got a bit confused and thought you were talking about the Ok(_) branch